ratis-server/src/main/java/org/apache/ratis/server/impl/MessageStreamRequests.java

```diff
@@ -39,25 +42,32 @@ private static class PendingStream {
     private final ClientInvocationId key;
     private long nextId = -1;
     private ByteString bytes = ByteString.EMPTY;
+    private List<ReferenceCountedObject<Message>> pendingRefs = new LinkedList<>();
 
     PendingStream(ClientInvocationId key) {
       this.key = key;
     }
 
-    synchronized CompletableFuture<ByteString> append(long messageId, Message message) {
+    synchronized CompletableFuture<ReferenceCountedObject<ByteString>> append(long messageId,
+        ReferenceCountedObject<Message> messageRef) {
       if (nextId == -1) {
         nextId = messageId;
       } else if (messageId != nextId) {
         return JavaUtils.completeExceptionally(new StreamException(
             "Unexpected message id in " + key + ": messageId = " + messageId + " != nextId = " + nextId));
       }
       nextId++;
+      Message message = messageRef.retain();
+      pendingRefs.add(messageRef);
       bytes = bytes.concat(message.getContent());
-      return CompletableFuture.completedFuture(bytes);
+
+      ReferenceCountedObject<ByteString> bytesRef = ReferenceCountedObject.delegateFrom(pendingRefs, bytes);
+      return CompletableFuture.completedFuture(bytesRef);
     }
 
-    synchronized CompletableFuture<ByteString> getBytes(long messageId, Message message) {
-      return append(messageId, message);
+    synchronized CompletableFuture<ReferenceCountedObject<ByteString>> getBytes(long messageId,
+        ReferenceCountedObject<Message> messageRef) {
+      return append(messageId, messageRef);
     }
```

Comment:
Make it return `CompletableFuture<ReferenceCountedObject<RaftClientRequest>>`.  The code will become simpler
```java
    synchronized CompletableFuture<ReferenceCountedObject<RaftClientRequest>> getWriteRequest(long messageId,
        ReferenceCountedObject<RaftClientRequest> requestRef) {
      return append(messageId, requestRef)
          .thenApply(bytes -> RaftClientRequest.toWriteRequest(requestRef.get(), () -> bytes))
          .thenApply(write -> ReferenceCountedObject.delegateFrom(pendingRefs, write));
    }
```